File: ServerSide/BallPassSchedulePattern/Players_QrcodeTypes/o_QRCodeTimingPatternPlayerDir/QRCodeTimingPatternPlayer.py
import os, sys
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class QRCodeTimingPatternPlayer(SuperPlayer):
    def __init__(self):
        super().__init__()  # スーパークラスの初期化メソッドを呼び出す
        self.my_name = None  # プレイヤーの名前
        self.qr_code_map = None  # QRコードのマップを保持する変数

    def return_my_name(self):
        return "QRCodeTimingPatternPlayer"

    def add_timing_pattern(self, qr_code_map):
        """
        バージョン2のQRコードにタイミングパターンを追加する関数
        入力: 2次元リスト (25x25)
        出力: タイミングパターンが追加された2次元リスト
        """
        grid_size = len(qr_code_map)
    
        # タイミングパターンは位置検出パターンの下と右にあるため、座標が(6,y)と(x,6)の位置に交互に配置
        for i in range(8, grid_size - 8):  # タイミングパターンは (6,y) と (x,6) の位置に配置
            # 横方向のタイミングパターン (y = 6)
            qr_code_map[6][i] = -2 if i % 2 == 0 else -1 # 偶数であれば-2(白色)、奇数であれば-1(黒色)とする。
            # 縦方向のタイミングパターン (x = 6)
            qr_code_map[i][6] = -2 if i % 2 == 0 else -1
    
        return qr_code_map


    def main(self):
        """
        このメソッド実行直前に、
        このスーパークラスのメンバ変数
        one_time_world_instanceに、
        最新のworldインスタンスを代入しています。
        
        このメソッド終了後、
        メインで使用しているworld_instanceを
        このスーパークラスのメンバ変数
        one_time_world_instanceで上書きし、
        更新しています。
        """
        
        # QRコードのタイミングパターンを追加
        self.qr_code_map = self.add_timing_pattern(self.one_time_world_instance.qr_code_map)

        print(f"{self.return_my_name()}が実行されました。")

        # 自身を更新
        self.one_time_world_instance.qRCodeTimingPatternPlayer = self

        return "Completed"

File: InstanceSide/BallPassSchedulePattern/Players_QRcodeTypes/o_QRCodeTimingPatternPlayerDir/QRCodeTimingPatternPlayer.cs
using UdonSharp;
using UnityEngine;

public class QRCodeTimingPatternPlayer : SuperPlayer
{
    [HideInInspector] public string myName;  // プレイヤーの名前
    [HideInInspector] public int version;  // QRコードのバージョン
    [HideInInspector] public int gridSize;  // グリッドサイズ (バージョンによる)
    [HideInInspector] public int[][] qrCodeMap;  // QRコードのマップ (2次元配列)

    public QRCodePositionDetectionPlayer qRCodePositionDetectionPlayer; // Unityエディタでアタッチ
    public RinaNumpy rinaNumpy; // UnityエディタでRinaNumpyをアタッチ

    private void QRCodeTimingPatternPlayerReset()
    {
        myName = "QRCodeTimingPatternPlayer";
        version = 0;
        gridSize = 0;
        qrCodeMap = null; // 初期化
    }

    public string ReturnMyName()
    {
        return "QRCodeTimingPatternPlayer";
    }

    private int[][] AddTimingPattern(int[][] qrCodeMap)
    {
        int gridSize = qrCodeMap.Length;

        // タイミングパターンの配置 (縦と横の交互配置)
        for (int i = 8; i < gridSize - 8; i++)
        {
            // 横方向のタイミングパターン (y = 6)
            qrCodeMap[6][i] = (i % 2 == 0) ? -2 : -1; // 偶数: -2 (白), 奇数: -1 (黒)
            // 縦方向のタイミングパターン (x = 6)
            qrCodeMap[i][6] = (i % 2 == 0) ? -2 : -1;
        }

        return qrCodeMap;
    }

    public override string ExecuteMain()
    {
        /*
         * メイン処理: タイミングパターンを追加
         */

        // QRコードの位置検出プレイヤーからデータを取得
        version = qRCodePositionDetectionPlayer.version;
        gridSize = version;
        qrCodeMap = qRCodePositionDetectionPlayer.qrCodeMap;

        // RinaNumpyでグリッドサイズチェック（例: 配列長の取得など）
        if (rinaNumpy == null)
        {
            Debug.LogError("RinaNumpy is not attached. Please attach it in the Unity Editor.");
            return "Failed";
        }

        // 必要な配列操作にRinaNumpyメソッドを活用
        float[] exampleArray = new float[gridSize]; // 必要に応じて変換
        for (int i = 0; i < exampleArray.Length; i++)
        {
            exampleArray[i] = i % 2 == 0 ? -2f : -1f;
        }

        // タイミングパターンを追加
        qrCodeMap = AddTimingPattern(qrCodeMap);

        // 実行完了メッセージ
        Debug.Log($"{ReturnMyName()}が実行されました。");

        // 必要な変数を再設定
        qRCodePositionDetectionPlayer.qrCodeMap = qrCodeMap;

        return "Completed";
    }
}

